src/open_pulse/pulse_utils.cpp

```diff
@@ -0,0 +1,122 @@
+#include "pulse_utils.hpp"
+#include "zspmv.hpp"
+
+complex_t internal_expect_psi_csr(const py::array_t<complex_t>& data,
+                                  const py::array_t<int>& ind,
+                                  const py::array_t<int>& ptr,
+                                  const py::array_t<complex_t>& vec) {
+    auto data_raw = data.unchecked<1>();
+    auto vec_raw = vec.unchecked<1>();
+    auto ind_raw = ind.unchecked<1>();
+    auto ptr_raw = ptr.unchecked<1>();
+
+    auto nrows = vec.shape(0);
+    complex_t temp, expt = 0;
+
+    for (decltype(nrows) row = 0; row < nrows; row++) {
+        temp = 0;
+        auto vec_conj = std::conj(vec_raw[row]);
+        for (auto jj = ptr_raw[row]; jj < ptr_raw[row + 1]; jj++) {
```

Comment:
nit: `jj` -> `j` ? 
I think it's less visual demanding... (very personal view though), but AFAIK we have been using single letters for loops through all the codebase, so for consistency I'd use one letter.

Reply:
I kept them from the original cython function, but I agree. I'll change it.